### PR TITLE
fix(vm,gqc): lower GQ_MIN_FRAME_DURATION to 5 ticks (20 FPS)

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Lowers the release animation-frame-duration floor from 20 ticks (5 FPS) to
5 ticks (20 FPS), per the owner's decision framework on
https://github.com/duplico/gamequeer/issues/304 ("the value gets picked
from measured worst case + margin").

## Measured basis

| Metric | Value |
|---|---|
| All-content delivered draw p100 (post-qc2024#16 event-driven SPI flush) | 40.67 ms |
| 5-tick budget (100 Hz tick, 50 ms/frame) | 50 ms |
| Margin at 5 ticks | 9.3 ms |
| 4-tick budget (40 ms/frame) | exceeded by the 40.67 ms p100 |
| Delivered cadence (post-#316 off-by-one fix) | exactly authored ticks |

gamequeer#304's most recent comment floated 4 ticks (25 FPS) as a
provisional target contingent on qc2024#16 landing masked-dither content at
~34-39 ms. The actual post-#16 measurement came in higher (40.67 ms),
which doesn't comfortably fit a 40 ms/4-tick budget but does fit the 50
ms/5-tick budget with real (if thin) margin -- so 5 ticks is the correct
pick under the issue's own "measured worst case + margin" rule, not the
literal "4 ticks" text of that comment. I've posted this reconciliation as
a comment on #304 (left open, per its own note that it stays open pending
the qc2024 submodule bump + on-badge confirmation).

## Changes

- `gamequeer/include/gamequeer.h`: `GQ_MIN_FRAME_DURATION` 20 -> 5, with a
  comment stating the current constraint (was previously undocumented).
- `gqc/src/gqc/datamodel.py` (QC2024-3 lockstep): the authoring-time
  frame-rate warning threshold moves from >5 FPS to >20 FPS, message text
  updated to match.
- `gqc/docs/authoring-and-perf-carts.md`: both mentions of the 20-tick/5
  FPS floor updated to 5-tick/20 FPS.
- `gamequeer/tests/golden/anim_advance.gq` + `gamequeer/tests/CMakeLists.txt`:
  retimed (see below).

## Golden fallout

Only `anim_advance.gq` (`frame_rate = 10`) is affected -- every other
golden-tested animation either has no `frame_rate` (default 5, i.e.
`ticks_per_frame = 20`, already >= both the old and new floor, unaffected
either way) or never advances a frame within its test's tick budget.
`anim_advance.gq`'s `frame_rate = 10` -> `ticks_per_frame = 10`, which was
below the old floor (20, clamped) but is now above the new floor (5,
unclamped) -- so its frames now run at their authored 10-tick duration
instead of the clamped 20.

Retimed `TICKS` (frame1: 21 -> 11, frame2: 41 -> 21) in
`tests/CMakeLists.txt`, and updated the fixture's header-comment timing
derivation to match. The committed golden PGMs did **not** need
regeneration -- confirmed pixel content is byte-identical at the new tick
offsets to what it was at the old ones (only the tick offset at which each
frame's content first appears moved), and confirmed the retimed goldens
genuinely fail (not a boundary coincidence -- TICKS=11 lands well inside
the pre-change code's frame-0 window, far from its 21-tick boundary)
against the pre-change 20-tick clamp.

## Verification

- Both flavors, clean Docker builds, full `ctest`:
  - `-DGQ_HEADLESS=ON`: 25/25 passed.
  - `-DGQ_HEADLESS=ON -DGQ_PERF_INSTRUMENT=ON -DGQ_PERF_INSTRUMENT_RUNS=ON`:
    30/30 passed.
- `gqc make test`: 0 tests collected (exit 5, expected -- the pytest suite
  is currently all TODO stubs; `pip install -e ".[test]"` succeeded with
  no import/syntax errors from the `datamodel.py` change).
- `clang-format -style=file --dry-run --Werror` on `gamequeer.h`: clean.
- Negative control: reverted `GQ_MIN_FRAME_DURATION` to 20 with the
  retimed `TICKS` values in place -- `golden_framebuffer_anim_advance_frame1`
  and `_frame2` both fail for real (genuine pixel mismatch, not a
  same-tick coincidence).

## Lockstep

QC2024-3: `gqc`'s warning threshold moved with the runtime clamp (see
`datamodel.py` diff above). No bytecode/struct format changes, so no
`gqc` compile-side changes beyond the warning message.

Related: gamequeer#304 (stays open), duplico/qc2024#16 (closed, basis for
the measurement), duplico/qc2024#45.

---
_AI-assisted (Claude Sonnet 5)._

### CODE REVIEW

**Comment by duplico** [comment]
## Adversarial review

The one-line clamp change is trivial; I focused on the lockstep-completeness sweep since that's where a change this small usually hides its risk.

### 1. Lockstep sweep — clean, nothing missed

`GQ_MIN_FRAME_DURATION` itself has exactly three consumers in the whole repo: the `#define` (`gamequeer.h`, updated), its use inside `gq_clamp_frame_duration()` (`gamequeer.c`, correctly parametric -- reads the macro, no baked `20`), and the doc mention (updated). Grepped for the literal old values too (`20 tick`, `5 FPS`, `200 ms`, `clamp`/`floor` language) across every `.c`/`.h`/`.py`/`.md`/`.gq`/CMake file in the repo: found exactly two false positives, both confirmed unrelated on inspection -- `authoring-and-perf-carts.md`'s "resampling at 5 FPS" line describes a specific committed GIF asset's ffmpeg resample rate (an authoring artifact, not the runtime floor), and `test_led_fade_delta.c`'s "5 (the minimum duration...)" is about the *LED cue* fade-interpolation threshold, a completely separate subsystem from animation frame ticks. No other stale numeric assumption found anywhere, including CI workflow files (no embedded tick/FPS assumptions there at all).

### 2. Retiming legitimacy — reproduced exactly, including the negative control

Confirmed `anim_advance.gq`'s authored `frame_rate = 10` directly from its source (`animations { pop <- "bwcircles.gif" { frame_rate = 10; ...` -- one line, unambiguous). Built the PR's code and swept `--ticks 1..30`, hashing every dump: boundaries land at **ticks 1, 11, 21** exactly, matching the PR's retimed `TICKS` values, and the three content hashes are byte-identical to the goldens *and* to the hashes I independently derived in the `#316` review two PRs ago (confirming pixel content really is unchanged, only the offset moved). Re-ran the negative control myself in a separate clone (no stash): reverted `GQ_MIN_FRAME_DURATION` back to 20 while keeping the retimed `TICKS=11/21` -- `frame1`/`frame2` both genuinely fail (frame0 still passes). Checked the boundary-coincidence risk explicitly: under the old floor, frame 0's window is `[1,20]`, and `TICKS=11` sits 9 ticks from either edge -- comfortably mid-window, not a coincidental near-boundary value that could pass by luck.

### 3. No other golden cart authors above the floor — checked every `.gq` myself, not trusted

Grepped every fixture under `tests/golden/` and `examples/games/perf/` for `frame_rate`: only `anim_advance.gq` (`frame_rate=10`) and the perf carts (`frame_rate=5`, not golden-referenced) specify one explicitly; everything else defaults to 5 FPS (`ticks_per_frame=20`). Since 20 ≥ both the old floor (20) and new floor (5), those fixtures' authored duration is genuinely unaffected either way -- confirmed this isn't a coincidence but a structural fact (the clamp only *raises* values strictly below the floor; 20 was never below either floor). Also confirmed `redundant_write.gq` has zero `animations{}` blocks at all, and every non-`anim_advance` fixture with an animation uses the CMake default `TICKS=5` (well under even the *new* 5-tick floor), so none of them can structurally reach a frame-advance boundary regardless of the clamp value.

### 4. gqc warning + pytest

Threshold (`>5` → `>20`) and message text both moved together in the same `datamodel.py` hunk, confirmed by direct diff read. Ran `pytest tests` myself: exit code 5, "collected 0 items" -- confirms the suite really is all-stub (pre-existing, `tests/test.py` is a literal TODO list, unrelated to this PR's `datamodel.py` change) and that the change introduced no import/syntax break. The PR correctly reports this rather than inventing test scope that doesn't fit the existing (nonexistent) test infrastructure.

### 5. `perf_draw_count_cross_check` at `TICKS=43` — structurally immune to the cadence change, not just currently passing

Read `run_perf_draw_count_cross_check.cmake` directly: this test does **not** assert any specific draw count. It asserts that two *independently-computed* counters (`--draw-count-out`'s `gq_draw_oled_stack_count` and `--perf-dump`'s `GQ_PERF_SEC_DRAW_OLED_STACK` count) **agree with each other**, whatever the actual number is -- both are incremented by the same `draw_oled_stack()` call, just via two different mechanisms, so they're guaranteed to move in lockstep regardless of how many times a 43-tick window happens to redraw under any cadence. The only other assertion (`draw_decode` zero/nonzero) is a coarse sufficiency check ("at least one RLE run happened"), trivially satisfied under either the old or new cadence at tick 43. Reproduced it passing on a clean build (`ctest -R perf_draw_count_cross_check`), but the more important finding is *why* it can't flake -- it was designed as a self-consistency check specifically so it wouldn't need to hardcode a cadence-dependent magic number.

### 6. Suites / CI / conventions / reconciliation comment

Clean builds, both flavors, matching CI exactly: `GQ_HEADLESS=ON` **25/25**, `+GQ_PERF_INSTRUMENT+RUNS` **30/30**. `clang-format --dry-run --Werror` on `gamequeer.h`: clean. No closing keywords anywhere (plain or quoted) in the commit message or PR body; `closingIssuesReferences` is empty `[]`, confirming `#304` stays open. Disclosure present. Pulled `gamequeer#304`'s full comment thread (6 comments) and confirmed the PR's reconciliation comment (2026-07-17T21:22:20Z) is accurate: it correctly states the measured 40.67 ms p100 (matching the PR body's own table exactly), correctly computes the 9.3 ms margin at 5 ticks (50 − 40.67 = 9.33), correctly explains why the *actual* measurement (higher than the provisional comment's 34-39 ms projection) means 5 ticks is the right pick under the issue's own stated "measured worst case + margin" rule rather than the stale literal "4 ticks" text from an earlier, projection-based comment, and correctly leaves the issue open pending the qc2024 submodule bump + on-badge confirmation.

### 7. `gamequeer.h` comment — current-truth only, no history narration

> "Floor on every animation frame's on-screen duration, in 100 Hz system ticks ... 5 ticks (50 ms/frame, 20 FPS) fits the render pipeline's measured worst-case delivered frame cost (~41 ms, all content) with margin to spare."

States the constraint and its measured justification in present tense; no "was previously 20" or "changed from X" narration anywhere in it. Fully compliant with the writing-rules standard applied consistently across this session's doc/comment reviews.

### Verdict

**Merge-ready.** This is a small, correctly-scoped change with a genuinely exhaustive lockstep sweep behind it -- I found nothing the PR missed. The retiming is reproduced exactly (same pixel hashes as my own independent `#316`-review sweep), the negative control is real (not a boundary coincidence), the cross-check test is provably cadence-independent rather than merely currently-passing, and the `#304` reconciliation comment accurately reflects the measured numbers driving the 5-vs-4-tick decision.

(AI-generated via Claude Code w/ Claude Sonnet 5)

File: gamequeer/tests/golden/anim_advance.gq [hunk 1]
```diff
@@ -19,47 +19,49 @@
...
- * `cmp` against dumps taken at the new TICKS values before/after the fix.
+ * three distinct hashes across the [1,10]/[11,20]/21+ tick ranges). The
+ * committed golden PGMs' *pixel content* is unaffected by lowering
+ * GQ_MIN_FRAME_DURATION (each frame's drawn bytes are identical to before,
+ * only the tick offset at which that content first appears moved),
+ * confirmed by byte-for-byte `cmp` against dumps taken at the new TICKS
+ * values before/after the change (gamequeer#304).
  *
  * Negative-control validation performed when authoring this fixture (per
  * the PR body): with the `frame_meta.data_pointer = 0` invalidation lines
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/tests/golden/anim_advance.gq:67]: Fixed in 745fc3d - updated the negative-control paragraph's stale TICKS=22/33 to the fixture's current TICKS=11/21 boundaries.

(AI-generated via Claude Code w/ Claude Sonnet 5)

File: gamequeer/include/gamequeer.h [hunk 2]
```diff
@@ -4,8 +4,14 @@
 #include <grlib.h>
 #include <stdint.h>
 
+// Floor on every animation frame's on-screen duration, in 100 Hz system
+// ticks (gq_clamp_frame_duration(), gamequeer.c) -- a `frame_rate` faster
+// than this many ticks/frame authors no differently than exactly this many.
+// 5 ticks (50 ms/frame, 20 FPS) fits the render pipeline's measured
+// worst-case delivered frame cost (~41 ms, all content) with margin to
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/include/gamequeer.h]: Fixed in 745fc3d - verified GQ_MIN_FRAME_DURATION is #ifndef-guarded (it already was; unchanged by this PR), so both halves apply: rewrote the comment to describe the clamp in ticks (FPS only as a parenthetical conversion), and moved the default's rationale inside the #ifndef block so it's clearly scoped to the default value rather than any override.

(AI-generated via Claude Code w/ Claude Sonnet 5)
